myst_nb/parser.py

```diff
@@ -40,53 +41,62 @@ def parse(self, inputstring, document):
         with renderer:
             # Loop through cells and render them
-            for cell in ntbk.cells:
+            for ii, cell in enumerate(ntbk.cells):
                 # Skip empty cells
                 if len(cell["source"]) == 0:
                     continue
-
-                # Cell container will wrap whatever is in the cell
-                classes = ["cell"]
-                for tag in cell.metadata.get("tags", []):
-                    classes.append(f"tag_{tag}")
-
-                # If a markdown cell, simply call the Myst parser and append children
-                if cell["cell_type"] == "markdown":
-                    source = cell["source"]
-                    if not source.endswith("\n"):
-                        # Ensure we always end with a newline in case it's a one-liner
-                        source = source + "\n"
-                    myst_ast = tokenize(source.splitlines(keepends=True))
-
-                    # Check for tag-specific behavior
-                    if "hide_input" in cell.metadata.get("tags", []):
-                        container = nodes.container()
-                        container["classes"].extend(["toggle"])
-                        with renderer.current_node_context(container, append=True):
-                            for child in myst_ast:
-                                renderer.render(child)
-                    else:
-                        for child in myst_ast:
-                            renderer.render(child)
-
-                # If a code cell, convert the code + outputs
-                elif cell["cell_type"] == "code":
-                    sphinx_cell = CellNode(classes=classes, cell_type=cell["cell_type"])
-                    cell_input = CellInputNode(classes=["cell_input"])
-                    sphinx_cell += cell_input
-                    renderer.current_node += sphinx_cell
-                    # Input block
-                    code_block = nodes.literal_block(text=cell["source"])
-                    cell_input += code_block
-
-                    # ==================
-                    # Cell output
-                    # ==================
-                    cell_output = CellOutputNode(classes=["cell_output"])
-                    sphinx_cell += cell_output
-
-                    outputs = CellOutputBundleNode(cell["outputs"])
-                    cell_output += outputs
+                try:
+                    _render_cell(cell, renderer)
+                except Exception as exc:
+                    doc2path = document.settings.env.doc2path
+                    path_page = Path(doc2path(document.settings.env.docname))
+                    srcdir = Path(document.settings.env.srcdir)
+                    relpath = path_page.relative_to(srcdir)
+                    raise Exception(
+                        f"Error parsing notebook {relpath}, cell #{ii+1}: {exc}"
+                    )
+
+
+def _render_cell(cell, renderer):
```

Comment:
none of the logic here is different (other than using `Document`), it's just moved into a function to make it easier to human-parse